### PR TITLE
Added ability to run sql queries on sqlite files

### PR DESCRIPTION
**Pull request by LesykMelnychuk**
Added ability to run execution and selection sql queries during sqlite db exploration.

### CODE REVIEW

**Comment by NSExceptional** [comment]
There's a lot of changes here. so when I have time, I'm going to review this on my machine and clean up the commit history a bit. I do want to merge this eventually though!

**Comment by LesykMelnychuk** [comment]
Great

**Comment by NSExceptional** [comment]
Hmmm, it appears you didn't check the option to give maintainers access to your fork when you created it. Could you give me commit access to your fork?

**Comment by NSExceptional** [comment]
Okay, I got everything squashed into a single commit on master and properly rebased.

After looking over your code, it doesn't match our coding conventions. Can you go through each of the files you've added or changed and make sure you're following the style set out in this codebase? Some examples:

- Do not indent `@property` declarations (I took care of this already)
- Never put spaces between arguments and their labels
   - ❌`[self foo:  bar];`
   - ✅`[self foo:bar];`
- Pointer markers go with the variable name, not the type
   - ❌`UILabel* textLabel;`
   - ✅`UILabel *textLabel;`
- There is no need to manually `@synthesize` properties unless you override the getter and the setter (so, use `self.isSelectionType` instead of `isSelectionType`)
- Prefer properties over ivars in almost all cases

**Comment by LesykMelnychuk** [comment]
Ok - I'll fix code style

File: Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m [hunk 1]
```diff
@@ -103,17 +101,76 @@ - (BOOL)close {
...
+    }
+
+    sqlite3_finalize(pstmt);
+
+    [self close];
+
+    return error;
+}
+
+- (NSArray<NSDictionary<NSString *, id> *> *)executeSelectionQuery: (NSString *)sql and: (NSString **)error {
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m]: Code style

File: Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m [hunk 2]
```diff
@@ -35,12 +36,90 @@ - (instancetype)initWithPath:(NSString *)path
...
-        [_dbm open];
-        [self getAllTables];
+        
+//        [self getAllTables];
     }
     return self;
 }
 
+- (void)viewWillAppear: (BOOL)animated {
+    [super viewWillAppear: animated];
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m]:
```suggestion
    [super viewWillAppear:animated];
```

File: Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m [hunk 3]
```diff
@@ -35,12 +36,90 @@ - (instancetype)initWithPath:(NSString *)path
...
+
+    UIAlertAction *cancelAction = [UIAlertAction
+        actionWithTitle: @"Cancel"
+        style: UIAlertActionStyleCancel
+        handler: nil
+    ];
+    
+    [alertController addAction: executeAction];
+    [alertController addAction: selectAction];
+    [alertController addAction: cancelAction];
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m]:
```suggestion
    [alertController addAction:cancelAction];
```

File: Classes/GlobalStateExplorers/SQLExecutionControllers/FLEXSQLCommandExecutionViewController.m [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    CGFloat textViewHeight = self.view.frame.size.height - startingY - submitButtonHeight - statusLabelHeight - sideMargin * 4;
+    
+    self.textView.frame = CGRectMake(sideMargin, startingY + sideMargin, width, textViewHeight);
+    self.statusLabel.frame = CGRectMake(sideMargin, self.textView.frame.origin.y + textViewHeight + sideMargin, width, statusLabelHeight);
+    self.submitButton.frame = CGRectMake(sideMargin, self.statusLabel.frame.origin.y + statusLabelHeight + sideMargin, width, submitButtonHeight);
+}
+
+- (void)submitPressed
+{
+    NSString *text = self.textView.text == nil ? @"" : self.textView.text;
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/SQLExecutionControllers/FLEXSQLCommandExecutionViewController.m]:
```suggestion
    NSString *text = self.textView.text ?: @"";
```

File: Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m [hunk 5]
```diff
@@ -103,17 +101,76 @@ - (BOOL)close {
...
+    
+    if (sqlite3_prepare_v2(_db, [sql UTF8String], -1, &pstmt, 0) == SQLITE_OK) {
+        while (sqlite3_step(pstmt) == SQLITE_ROW) {
+            NSUInteger num_cols = (NSUInteger)sqlite3_data_count(pstmt);
+            
+            if (num_cols > 0) {
+                NSMutableDictionary<NSString *, id> *dict = [NSMutableDictionary dictionaryWithCapacity:num_cols];
+                
+                int columnCount = sqlite3_column_count(pstmt);
+
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m]: Avoid excess whitespace, group lines together logically

File: Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m [hunk 6]
```diff
@@ -35,12 +36,90 @@ - (instancetype)initWithPath:(NSString *)path
...
+        alertControllerWithTitle: @"Select you option"
+        message: nil
+        preferredStyle: UIAlertControllerStyleActionSheet
+    ];
+    
+    UIAlertAction *executeAction = [UIAlertAction
+        actionWithTitle: @"Execute SQL"
+        style: UIAlertActionStyleDefault
+        handler: ^(UIAlertAction *action) {
+            FLEXSQLCommandExecutionViewController* controller = [FLEXSQLCommandExecutionViewController new];
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/DatabaseBrowser/FLEXTableListViewController.m]:
```suggestion
            FLEXSQLCommandExecutionViewController *controller = [FLEXSQLCommandExecutionViewController new];
```

There are still a lot of these

File: Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m [hunk 7]
```diff
@@ -103,17 +101,76 @@ - (BOOL)close {
...
+                
+                int columnCount = sqlite3_column_count(pstmt);
+                
+                int columnIdx = 0;
+                
+                for (columnIdx = 0; columnIdx < columnCount; columnIdx++) {
+                    NSString *columnName = [NSString stringWithUTF8String:sqlite3_column_name(pstmt, columnIdx)];
+                    id objectValue = [self objectForColumnIndex:columnIdx stmt:pstmt];
+                    
+                    [dict setObject:objectValue forKey:columnName];
```
Comment (NSExceptional) [inline Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m]:
```suggestion
                    dict[columnName] = objectValue;
```

Use subscripting syntax wherever possible

**Comment by NSExceptional** [comment]
I left some comments, there's a lot you missed 😅 I didn't mark them all, just a bunch I found.

Also, before you go any further, make sure your HEAD is at `origin/master` to avoid the merge commit. You don't need to merge, just fast forward. Lemme know if you need help with avoiding a merge commit.

**Comment by LesykMelnychuk** [comment]
> I left some comments, there's a lot you missed 😅 I didn't mark them all, just a bunch I found.
> 
> Also, before you go any further, make sure your HEAD is at `origin/master` to avoid the merge commit. You don't need to merge, just fast forward. Lemme know if you need help with avoiding a merge commit.

Fixed comments you made - fixed, were could found.
Thanks, that you are interested in my pull request!

**Comment by NSExceptional** [comment]
@revolter Can you tell me what this macro and code segment is for? 😅I'm trying to decipher some of this old code, I've never used the sqlite APIs

https://github.com/Flipboard/FLEX/blame/0de3a9d65c54533848b6721d0d66bc3fb0d740fa/Classes/GlobalStateExplorers/DatabaseBrowser/FLEXSQLiteDatabaseManager.m#L38

**Comment by revolter** [comment]
That macro is set by the SQLCipher pod and the code is used to decrypt a crypted database after opening it, or else any query would fail.

**Comment by NSExceptional** [comment]
Thanks! I'm trying to add comments to these files

**Comment by NSExceptional** [comment]
@LesykMelnychuk Can I ask why you added `executeNonSelectQuery:` and `executeSelectionQuery:error:` instead of making use of `executeQuery:`?

I've read the code more thoroughly now and there seems to be a lot of duplication now. At the very least, would you be able to make those two methods call into `executeQuery:`?

**Comment by LesykMelnychuk** [comment]
@NSExceptional Hi, cause `executeSelectionQuery:error` returns result of selection

**Comment by NSExceptional** [comment]
Your `executeSelectionQuery:error:` is almost identical to the original `executeQuery:` unless I'm overlooking something. You return an error in an argument but that's the only difference

**Comment by LesykMelnychuk** [comment]
> Your `executeSelectionQuery:error:` is almost identical to the original `executeQuery:` unless I'm overlooking something. You return an error in an argument but that's the only difference

Understud, - fixed.

**Comment by LesykMelnychuk** [comment]
Hello @NSExceptional , can you please merge this pull ?

**Comment by NSExceptional** [comment]
Hey @LesykMelnychuk,

This PR isn't in a state where I'm ready to merge it. I have a fork of your branch I started working on a while back but I got sidetracked and it fell off my todo list. I will try to get it done and merged this week. (At the moment I don't even remember what I was working on specifically, probably adding comments or something.)

**Comment by revolter** [comment]
Shouldn't I fix the conflicts in #305, modify things if needed, merge it, and then rebase this, instead of having https://github.com/Flipboard/FLEX/pull/294/commits/f7659a8910dd1fc685f7ea180bdf9ea22719e905 here?

**Comment by NSExceptional** [comment]
@revolter @LesykMelnychuk I'm currently in the process of rebasing and resolving conflicts between this branch and the 4.0 branch, should have it done tomorrow sometime

**Comment by NSExceptional** [comment]
Accidentally named the branch `4.0-flex` when I meant to just name it `4.0`, so I renamed it just now, FYI

**Comment by revolter** [comment]
What I meant was that this PR included #305, so you merged 2 things at the same time, while #305 looks now like it wasn't merged.

**Comment by NSExceptional** [comment]
Are you saying @LesykMelnychuk somehow included #305 in this original PR, and that I unknowingly discarded it when I made my changes and merged this just now?

**Comment by revolter** [comment]
Oooh, it was probably incorrectly rebased. Not it's ok, sorry I didn't check again, today. So it's all good 👍

**Comment by NSExceptional** [comment]
I'm so confused :P What's wrong?

**Comment by revolter** [comment]
I am what's wrong 😆 Sorry for the confusion, and completely disregard this conversation. As I said, everything is ok 😃

**Pull request merged**